architecture/index.md

```diff
@@ -39,6 +39,55 @@ API resources in the second row (PipelineRun, Snapshot) should primarilly be tho
 data-plane resources. The system responds to user requests by creating and managing the lifecycle of
 these resources.
 
+- An [Application] represents a functionally coherent set of [Components] that should be built,
+  tested, and released together. The user provides and names their [Applications]. They are
+  generally long-lived and don't change much after they are created.
+- A [Component] represents an OCI artifact to build from a particular git branch (and a particular context
+  directory). It is owned by an [Application]. The user provides and names their [Components] and
+  specifies which git repositories and branches define how the artifact is built. The user may add,
+  remove, or change [Components] over the lifespan of an [Application].
+- A [Snapshot] represents a collection of particular OCI artifact, mapped to the [Components] of an
+  [Application]. A [Snapshot] is owned by an [Application]. Generally, the [integration-service]
+  produces new [Snapshots] automatically in response to completed builds. It is also possible for
+  the user to construct and provide [Snapshots] to test or release deliberate combinations of
+  OCI artifacts. Old [Snapshots] are garbage collected if not bound to other resources like
+  a [SnapshotEnvironmentBinding] or a [Release].
+- An [IntegrationTestScenario] represents a test that should be run against new [Snapshots] that
+  belong to a particular [Application]. It is owned by an [Application]. The user provides writes
+  tekton pipelines that test their application and registers them with the system by creating
+  [IntegrationTestScenarios].
+- An [Environment] represents a destination that [application-service] and the [gitops-service] can
+  deploy to. Multiple [Applications] in the same workspace can deploy to the same
+  [Environment]. An [Environment] may be long lived (like the `development` [Environment] provided
+  by the system as part of the initialization of a user's workspace) or may be short-lived and
+  ephemeral (like the temporary [Environments] created by the [integration-service] as a part of its
+  test execution processes).
+- A [DeploymentTargetClaim] represents a request to provision a new deployment target. An
+  [Environment] is backed by a [DeploymentTargetClaim]. The [DeploymentTargetClaim] causes new
+  deployment targets to be provisioned which are represented by [DeploymentTargets] that get bound
```

Comment:
How are Environment and DeploymentTarget related? It seems like Environment is backed by DeploymentTargetClaim, but it also seems like DeploymentTarget is a resolved / instantiated DeploymentTargetClaim. Is DeploymentTarget then the infrastructural detail where an Environment is made available?

Reply:
You got it. The DeploymentTarget represents the concrete instantiated remote resource. The Environment is backed by the claim, and the claim is *bound* to the DeploymentTarget. Only one DeploymentTarget can be claimed at a time (said another way, only one DeploymentTargetClaim can be bound to a DeploymentTarget at a time). The DeploymentTargetClaim permits managing exclusive access to DeploymentTargets.

It is meant to "look like" the PersistentVolumeClaim API. We modelled it off that. In the PVC APIs, you can have a PersistentVolume that lingers and is claimed multiple times over its life. The PV represents the real storage resource somewhere. You can also have ephemeral PVs that come into and out of existence in response to PVCs being created, say provisioned via AWS APIs. Same idea with the DTC/DT APIs.